File: ServidorPrincipal/src/main/java/com/jbrod/servidorprincipal/trivias/Trivia.java

package com.jbrod.servidorprincipal.trivias;

import java.util.LinkedList;

/**
 *
 * @author jorge
 */
public class Trivia {
    private String id_trivia;
    private String nombre;
    private String usuario_creador; 
    private String fecha_creacion; 
    private int tiempoPregunta; 
    private String tema; 
    
    private LinkedList <Componente> componentes;

    
    
    
    /**
     * Crea una trivia en base a:
     * @param id: String con el identificador que tendra la trivia.
     * @param nombre: String con el nombre de la trivia.
     * @param tema: String con el tema de la trivia.
     * @param user: String con el identificador del usuario creador.
     * @param date: String con la fecha de creación de la trivia.
     * @param tiempo: int con el tiempo de cada pregunta.
     **/
    public Trivia(String id, String nombre, String tema, String user, String date, int tiempo){
        this.tema = tema;
        this.nombre = nombre;
        id_trivia = id; 
        usuario_creador = user; 
        fecha_creacion = date; 
        this.tiempoPregunta = tiempo;
        
        componentes = new LinkedList<>();
    }
    
    
    
    public String eliminarComponente(String id){
        for (Componente componente : componentes) {
            if(id.equals(componente.getId_componente())){
                componentes.remove(componente);
                return "El componente " + id + " ha sido eliminado correctamente.";
            }
        }
        return "No se encontro el componente " + id + " dentro de " + this.id_trivia + " para eliminar";
    }
    
    public String agregarComponente(Componente componente){
        if(componente != null){
            String nm = componente.getId_componente();
            for (Componente c : componentes) {
                if(nm.equals(c.getId_componente())){
                    return "No se puede agregar el componente " + nm + " en " + id_trivia + " porque ya existe un componente con ese id.";
                }
            }
            componentes.add(componente);
            return "El componente con id " + nm + " se ha agregado correctamente a " + id_trivia + ".";
        }else{
            return "No se puede agregar el componente en " + id_trivia + " porque es nulo.";
        }
    }
    
    public String modificarComponente(Componente componente){
        if(componente != null){
            String nm = componente.getId_componente();
            for (Componente c : componentes) {
                if(nm.equals(c.getId_componente())){
                    
                    componentes.add(componentes.indexOf(c), componente);
                    componentes.remove(c);
                    return "El componente " + nm + " se modifico correctamente en " + id_trivia + ".";
                }
            }
            return "No se pudo actualizar el componente " + nm + " en " + id_trivia + " porque no existe ningun componente con ese id.";
        }else{
            return "No se puede modificar el componente en " + id_trivia + " porque es nulo.";
        }        
    }
    
    
    public String exportarTrivia(){
        String trivia = "trivia(\n";
            trivia += "    \"ID_TRIVIA\" : " + id_trivia       + "\n";
            trivia += "    \"NOMBRE\"    : " + nombre          + "\n";
            trivia += "    \"CREADOR\"   : " + usuario_creador + "\n";
            trivia += "    \"FECHA\"     : " + fecha_creacion  + "\n";
            trivia += "    \"TIEMPO\"    : " + tiempoPregunta  + "\n";
            trivia += "    \"TEMA\"      : " + tema            + "\n";
            trivia += "\n";
        
        for (Componente componente : componentes) {
            if(componente != null){
                trivia += componente.obtenerCodigo() + "\n";
            }
        }
        
        trivia += ")";
        return trivia;
        
    }
    
    
    

    public String getNombre() {
        return nombre;
    }

    public void setNombre(String nombre) {
        this.nombre = nombre;
    }

    public String getTema() {
        return tema;
    }

    public void setTema(String tema) {
        this.tema = tema;
    }    
    
    public int getTiempoPregunta() {
        return tiempoPregunta;
    }

    public void setTiempoPregunta(int tiempoPregunta) {
        this.tiempoPregunta = tiempoPregunta;
    }
    
    public String getId_trivia() {
        return id_trivia;
    }

    public void setId_trivia(String id_trivia) {
        this.id_trivia = id_trivia;
    }

    public String getUsuario_creador() {
        return usuario_creador;
    }

    public void setUsuario_creador(String usuario_creador) {
        this.usuario_creador = usuario_creador;
    }

    public String getFecha_creacion() {
        return fecha_creacion;
    }

    public void setFecha_creacion(String fecha_creacion) {
        this.fecha_creacion = fecha_creacion;
    }

    public LinkedList<Componente> getComponentes() {
        return componentes;
    }

    public void setComponentes(LinkedList<Componente> componentes) {
        this.componentes = componentes;
    }
}

File: AppTrivias/app/src/main/java/trivias/Trivia.kt
package trivias

import java.util.LinkedList

class Trivia(
    private var id_trivia: String,
    private var nombre: String,
    private var usuario_creador: String,
    private var fecha_creacion: String,
    private var tiempo_pregunta: Int,
    private var tema: String
    //Agregar la lista de componentes
) {



    public var componentes: LinkedList<Componente>
    public var actual:Int = 0

    // Puedes agregar un bloque init si necesitas realizar alguna acción al crear la instancia
    init {
        println("Trivia creada: $nombre, ID: $id_trivia")
        componentes = LinkedList()
    }


    fun avanzar(){
        if(actual < componentes.size ){
            actual ++
            println("Actual: " + actual)
        }
    }

    fun obtenerActual():Componente{
        return componentes.get(actual)
    }


    fun agregarComponente(c:Componente){
        componentes.add(c)
    }



    // Getters
    fun getIdTrivia(): String {
        return id_trivia
    }

    fun getNombre(): String {
        return nombre
    }

    fun getUsuarioCreador(): String {
        return usuario_creador
    }

    fun getFechaCreacion(): String {
        return fecha_creacion
    }

    fun getTiempoPregunta(): Int {
        return tiempo_pregunta
    }

    fun getTema(): String {
        return tema
    }

    fun mostrarRespuestas(): String{
        var r = ""

        for (elemento in componentes){
            if(elemento != null){
                r += elemento.texto_visible + "\n" + elemento.respuestaUsuario + "\n\n"
            }
        }
        return r
    }

    fun obtenerPuntuacion(us:String, tp:Int):String{
        var p = ""

        var puntuacion = 0
        for(componente in componentes){
            puntuacion += componente.obtenerPuntuacion()
        }
        //Regla de 3 para dimensionar la puntuacion
        puntuacion = puntuacion * 100 / (componentes.size * 100)

        p += "<!realizar_solicitud: \"AGREGAR_PUNTUACION\">\n"
        p += "{  \"PARAMETROS_PUNTUACION\":[{\n"
        p += "    \"TRIVIA\": "     + id_trivia             + "\n"
        p += "    \"USUARIO\": "    + us                    + "\n"
        p += "    \"TIEMPO\": "     + tp.toString()         + "\n"
        p += "    \"PUNTUACION\": " + puntuacion.toString() + "\n"
        p += "}\n"
        p += "]}\n"
        p += "<fin_solicitud_realizada!>"

        return p
    }


}
